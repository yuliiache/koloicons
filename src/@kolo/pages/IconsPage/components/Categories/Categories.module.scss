@import "src/@kolo/design";

.categoryList {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  background-color: $LightBG;
  padding: 16px;
  list-style: none;
  transition: all .25s linear;
  height: calc(100vh - 162px);
  overflow-y: auto;

  &::-webkit-scrollbar {
    width: 4px;
  }

  &::-webkit-scrollbar-thumb {
    width: 4px;
    opacity: .5;
    border-radius: 4px;
    background-color: $ScrollbarColor;
  }
}

.item {
  width: 100%;

  &:hover {
    border-radius: 4px;
    background-color: $DarkBG;
    color: $secFontColor;
    transition: all .2s linear;
  }
}
