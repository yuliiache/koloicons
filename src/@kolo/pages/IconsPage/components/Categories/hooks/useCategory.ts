import {useState} from 'react';
import {useSearchParams} from 'react-router-dom';

interface useCategoryReturn {
  activeCategoryId: string | null;
  handleCategoryClick: (value: string) => void;
}

const useCategory = (): useCategoryReturn => {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const [searchParams, setSearchParams] = useSearchParams();
  const [activeCategoryId, setActiveCategoryId] = useState<string | null>(null);

  const handleCategoryClick = (id: string) => {
    setSearchParams({category: id});
    setActiveCategoryId(id);
  };

  return {activeCategoryId, handleCategoryClick};
};

export default useCategory;
