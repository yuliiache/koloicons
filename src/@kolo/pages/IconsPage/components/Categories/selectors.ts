import {RootState} from 'app/store';
import {createSelector, Selector} from 'reselect';

import {CATEGORIES_KEY, CategoriesState} from './slice/categoriesSlice';

export const getIconCategories: Selector<RootState, CategoriesState> = (state) => state[CATEGORIES_KEY];

export const iconCategoriesSelector: Selector<RootState, CategoriesState> = createSelector(
  getIconCategories,
  (categories: CategoriesState): CategoriesState => ({
    categoriesList: categories.categoriesList,
  })
);
