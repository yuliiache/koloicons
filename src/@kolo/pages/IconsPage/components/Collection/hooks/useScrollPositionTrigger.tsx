import {useEffect, useState} from 'react';

export const useScrollPositionTrigger = () => {
  const [isAbsolutePosition, setIsAbsolutePosition] = useState(false);

  useEffect(() => {
    function updatePosition() {
      setIsAbsolutePosition(window.scrollY >= window.innerHeight / 6); //Temporary solution
    }

    window.addEventListener('scroll', updatePosition);
    updatePosition();

    return () => window.removeEventListener('scroll', updatePosition);
  }, []);

  return isAbsolutePosition;
};
