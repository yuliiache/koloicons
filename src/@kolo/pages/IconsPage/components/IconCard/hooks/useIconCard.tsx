import {useState} from 'react';

interface UseIconCardReturn {
  isAdded: boolean;
  handleClick: () => void;
}

const useIconCard = (): UseIconCardReturn => {
  const [isAdded, setIsAdded] = useState(false);
  const handleClick = () => {
    setIsAdded(!isAdded);
  };

  return {isAdded, handleClick};
};

export default useIconCard;
