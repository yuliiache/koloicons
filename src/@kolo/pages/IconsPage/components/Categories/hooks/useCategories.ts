import {useAppSelector} from 'app/hooks';

import {iconCategoriesSelector} from '../selectors';
import {CategoriesData} from '../slice/categoriesSlice';

interface CategoriesHookReturn {
  categoriesList: CategoriesData[];
}

const useCategories = (): CategoriesHookReturn => {
  const {categoriesList} = useAppSelector(iconCategoriesSelector);

  return {categoriesList};
};

export default useCategories;
