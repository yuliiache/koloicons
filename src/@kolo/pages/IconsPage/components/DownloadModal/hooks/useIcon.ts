import {useEffect} from 'react';
import {useDispatch, useSelector} from 'react-redux';

import {getIconStart} from '../iconSlice';
import {IconSelector} from '../selectors';

const useIcon = (iconId: string) => {
  const icon = useSelector(IconSelector);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(getIconStart(iconId));
  }, [dispatch, iconId]);

  return {icon};
};

export default useIcon;
