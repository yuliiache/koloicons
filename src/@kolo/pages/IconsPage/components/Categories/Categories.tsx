import {FC} from 'react';

import styles from './Categories.module.scss';
import CategoryButton from './CategoryButton/CategoryButton';
import useCategories from './hooks/useCategories';
import useCategory from './hooks/useCategory';
import {CategoriesData} from './slice/categoriesSlice';

const {categoryList, item} = styles;

const Categories: FC = () => {
  const {categoriesList} = useCategories();
  const {activeCategoryId, handleCategoryClick} = useCategory();

  return (
    <ul className={categoryList}>
      {categoriesList.map(({name, id}: CategoriesData) => (
        <li
          key={id}
          className={item}
        >
          <CategoryButton
            name={name}
            onClick={() => handleCategoryClick(id)}
            isActive={id === activeCategoryId}
          />
        </li>
      ))}
    </ul>
  );
};

export default Categories;
