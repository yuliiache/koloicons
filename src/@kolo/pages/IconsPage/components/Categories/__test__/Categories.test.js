import React from 'react';
import {fireEvent, render, screen} from '@testing-library/react';

import '@testing-library/jest-dom';

import Categories from '../Categories';
import * as useCategoriesHook from '../hooks/useCategories';
import * as useCategoryHook from '../hooks/useCategory';

jest.mock('../CategoryButton/CategoryButton', () => {
  const MockCategoryButton = ({name, onClick, isActive}) => (
    <button
      data-testid="category-button"
      data-active={isActive.toString()}
      onClick={onClick}
    >
      {name}
    </button>
  );

  MockCategoryButton.displayName = 'MockCategoryButton';

  return MockCategoryButton;
});

describe('Categories Component', () => {
  beforeEach(() => {
    jest.spyOn(useCategoriesHook, 'default').mockReturnValue({
      categoriesList: [],
      isLoading: true,
    });
    jest.spyOn(useCategoryHook, 'default').mockReturnValue({
      activeCategoryId: null,
      handleCategoryClick: jest.fn(),
    });
  });

  it('renders null when isLoading is true', () => {
    render(<Categories />);
    const categoryList = screen.queryByTestId('category-list');
    expect(categoryList).not.toBeInTheDocument();
  });

  it('renders categories list correctly and matches snapshot', () => {
    useCategoriesHook.default.mockReturnValue({
      categoriesList: [
        {id: '1', name: 'Category 1'},
        {id: '2', name: 'Category 2'},
      ],
      isLoading: false,
    });
    render(<Categories />);
    const buttons = screen.getAllByTestId('category-button');
    expect(buttons.length).toBe(2);
    expect(buttons[0]).toHaveTextContent('Category 1');
    expect(buttons[1]).toHaveTextContent('Category 2');
  });

  it('correctly handles category click', () => {
    const handleCategoryClickMock = jest.fn();
    useCategoriesHook.default.mockReturnValue({
      categoriesList: [{id: '1', name: 'Category 1'}],
      isLoading: false,
    });
    useCategoryHook.default.mockReturnValue({
      activeCategoryId: '1',
      handleCategoryClick: handleCategoryClickMock,
    });
    render(<Categories />);
    const button = screen.getByText('Category 1');
    fireEvent.click(button);
    expect(handleCategoryClickMock).toHaveBeenCalledWith('1');
  });

  it('matches snapshot with categories list', () => {
    useCategoriesHook.default.mockReturnValue({
      categoriesList: [
        {id: '1', name: 'Category 1'},
        {id: '2', name: 'Category 2'},
      ],
      isLoading: false,
    });
    const {asFragment} = render(<Categories />);
    expect(asFragment()).toMatchSnapshot();
  });
});
