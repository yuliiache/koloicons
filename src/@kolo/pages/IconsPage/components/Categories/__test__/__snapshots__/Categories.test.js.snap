// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories Component matches snapshot with categories list 1`] = `
<DocumentFragment>
  <ul
    class="categoryList"
  >
    <li
      class="item"
    >
      <button
        data-active="false"
        data-testid="category-button"
      >
        Category 1
      </button>
    </li>
    <li
      class="item"
    >
      <button
        data-active="false"
        data-testid="category-button"
      >
        Category 2
      </button>
    </li>
  </ul>
</DocumentFragment>
`;
